### ISSUE TITLE
Ampache XML can be mal-formed - includes characters invalid in XML

### ISSUE THREAD

**Issue by jonaseberle** [owncloud/music#723]
For my music collection, the output XML is invalid. 

In my case, these are control characters x01 and x04 which are not valid in XML:
```
<song id='9250'>
		<title>)/</title>
...
</song>
```

The "music" web interface handles these titles somehow:
![image](https://user-images.githubusercontent.com/1678001/67633239-b9198980-f8ad-11e9-9737-1f8570a078fd.png)

I suggest hardening the XML output against such stuff.

**Comment by jonaseberle** [comment]
This was tested in music 0.11.0

I suggest to use an XML-library in the templates e.g. https://github.com/owncloud/music/blob/master/templates/ampache/songs.php 

Changing the code to 
```
<?php
print '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>';
$x = new DOMDocument();
?>
<root>
<?php foreach ($_['songs'] as $song): ?>
        <song id='<?php p($song->getId());?>'>
                <title><?php p($x->saveXML($x->createTextNode($song->getTitle())));?></title>
...
```
does fix it.

Of course it would be even more sane to just use DomDocument outright to create the XML. IMHO that would have less code smell, too.

If you want, I can provide a PR for that.

**Comment by paulijar** [comment]
Thanks for the report.

Have you happened to test the Subsonic API on the same audio file; is the XML malformed there, too? No specific steps are taken on the Subsonic API to sanitize the strings for XML, but the XML is built with different APIs, so it could happen implicitly. 

I wouldn't try to sanitize the strings within the XML templates because then the same `$x->saveXML($x->createTextNode($someString))` mantra would need to be repeated a zillion times. It might make most sense to do it already when scanning the library and extracting a metadata tag here: https://github.com/owncloud/music/blob/0c48f2097ec9ba867204e8316ff51697b9079454/utility/extractorgetid3.php#L97. There, we already clean up any invalid UTF-8 characters and the control characters could be dropped at the same point. The downside would be that this would require rescanning your library before the fix would take effect.

Furthermore, it might be cleaner to do something like [this ](https://stackoverflow.com/a/1497928) instead of converting the string first to DOM and then back to string.

**Comment by jonaseberle** [comment]
The `$x->saveXML($x->createTextNode($someString))` was just an example.

IMHO output needs to be sanitized while outputting. Other means could be done additionally.

I would rather use DOMDocument's API completely to build the XML.

Then, at the end of the template just `echo $x->saveXML()` or (I did not look up what p() is doing yet but I see it used to output) `p($x->saveXML())`.

I volunteer to do that for ALL the templates.

**Comment by paulijar** [comment]
Sorry for the delay, I've been extra busy lately with other duties.

One thing to consider is that the purpose of the `TemplateResponse` mechanism is to provide declarative means to describe the contents of a HTTP response body. Meanwhile, the DOMDocument API provides imperative means to build XML, so it's a somewhat poor match. If the response was built with DOMDocument, then this logic probably shouldn't reside under the `templates` folder and instead of `TemplateResponse` type e.g. `DataDisplayResponse` could be used.

**Comment by jonaseberle** [comment]
Sorry, I am not much into the inner workings of a nextcloud/owncloud extension. 
Maybe I did not understand completely but I would say it would be architecturally equally good as it is now, so-to-say "template generates XML". Just with a library that takes care of W3C-validity during output.

**Comment by paulijar** [comment]
Generally speaking, "template" (*) is something which resembles the end result, but with some blanks which should be filled in later. And this is kind of how the Ampache XML templates work: They look like XML documents, but some parts of the documents are bound later from the passed in parameters.

It may be equally good architecture to use a library to create the XML in pure PHP syntax. My point is just, that if it's not a template, then IMHO it shouldn't be called a template.

If the responses would be built using the DOMDocument, then one alternative could be to place this logic within the namespace OCA\Music\Http. There could be classes like `AmpahceAlbumsResponse`, `AmpachePlaylistsResponse`, etc. which would be responsible of building the corresponding response type.

(*) Dictionary definition: A generic model or pattern from which other objects are based or derived.

**Comment by paulijar** [comment]
I ended up refactoring `AmpacheController` so that it doesn't use any templates. Instead, responses are built the same way as in `SubsonicController`, by utilizing our `XMLResponse` class. This, on the other hand, uses `DOMDocument` to build the XML document so this kind of is the solution you suggested.

This resulted in slightly more code in `AmpacheController` but I feel its logic is now more straightforward. The downside is that it isn't now quite so easy to see what the resulting XML will look like, but this isn't a significant issue.

**Comment by jonaseberle** [comment]
Splendid, thank you very much!

**Comment by jonaseberle** [comment]
Sorry I was not able to test earlier. It is fixed in music 0.11.1. 

Thank you!

**Comment by paulijar** [comment]
Hmm, it shouldn't be yet fixed in 0.11.1 because I made the fix after releasing that version 🤔. So you should see the fix only if you are using the most recent version downloaded from github.

**Comment by jonaseberle** [comment]
Lol, then it was just coincidence. I hope I'll find the time to retest.

**Comment by jonaseberle** [comment]
I tested the current master with nextcloud 16.06.

Dates are differently encoded now. I think they should be ISO 8601 (https://github.com/ampache/ampache/wiki/XML-API). By looking at older code I see `date('c')` was used there.

rhythmbox's ampache client does not start downloading the _songs playlist now as it does with 0.11.1.

Music 0.11.1:
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?><root>
	<auth>...</auth>
	<version>350001</version>
	<update>2019-12-08T14:31:46+00:00</update>
	<add>2019-12-08T14:31:46+00:00</add>
	<clean>2019-12-08T14:31:46+00:00</clean>
	<songs>6635</songs>
	<artists>792</artists>
	<albums>1277</albums>
	<playlists>2</playlists>
	<session_expire>2019-12-08T16:11:46+00:00</session_expire>
	<tags>0</tags>
	<videos>0</videos>
</root>
```

Music master:
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?><root>
	<auth>...</auth>
	<version>350001</version>
	<update>1575815426</update>
	<add>1575815426</add>
	<clean>1575815426</clean>
	<songs>6635</songs>
	<artists>792</artists>
	<albums>1277</albums>
	<playlists>2</playlists>
	<session_expire>2019-12-08T16:10:26+00:00</session_expire>
	<tags>0</tags>
	<videos>0</videos>
</root>
```

Thank you for your work!

**Comment by jonaseberle** [comment]
It might be better if I put that comment to the commit. Because the XML-encoding from this issue is actually fixed.

**Comment by paulijar** [comment]
Fixes for the original issue, and the update/add/clean date issue were released in v0.12.0, which was just hotfixed with v0.12.1 because of a fatal UI issue when running on Nextcloud 14+.
